### PR TITLE
Update calico-node to use Calico V2 API

### PR DESCRIPTION
**Pull request by gunjan5**
## Description
Update Calico node startup to use the V2 APIs

## Todos

- [ ] Tests
- [ ] Documentation
- [ ] Release note

## Release Note

```release-note
None required
```

### CODE REVIEW

File: calico_node/startup/startup.go [hunk 1]
```diff
@@ -249,7 +272,7 @@ func configureIPsAndSubnets(node *api.Node) bool {
 	// set of configuration as it makes the processing below easier, and we
 	// must end up configuring some BGP fields before we complete.
 	if node.Spec.BGP == nil {
-		log.Info("Initialise BGP data")
+		log.Info("Initialize BGP data")
```
Comment (caseydavenport) [inline calico_node/startup/startup.go]: 👍

**Comment by gunjan5** [comment]
closing and reopening because Semaphore disappeared

File: calico_node/Makefile [hunk 2]
```diff
@@ -469,6 +470,8 @@ node-test-containerized: vendor run-etcd-host run-k8s-apiserver
 	-v $(VERSIONS_FILE):/versions.yaml:ro \
 	-e VERSIONS_FILE=/versions.yaml \
 	-e LOCAL_USER_ID=$(LOCAL_USER_ID) \
+	-e DATASTORE_TYPE=$(DATASTORE_TYPE) \
+    -e ETCD_ENDPOINTS=http://$(LOCAL_IP_ENV):2379 \
```
Comment (robbrockbank) [inline calico_node/Makefile]: should be tab not spaces.

File: calico_node/Makefile [hunk 3]
```diff
@@ -469,6 +470,8 @@ node-test-containerized: vendor run-etcd-host run-k8s-apiserver
 	-v $(VERSIONS_FILE):/versions.yaml:ro \
 	-e VERSIONS_FILE=/versions.yaml \
 	-e LOCAL_USER_ID=$(LOCAL_USER_ID) \
+	-e DATASTORE_TYPE=$(DATASTORE_TYPE) \
```
Comment (robbrockbank) [inline calico_node/Makefile]: does it make sense parameterizing this here - since we also pass in the etcd endpoints?  i.e. if you change the type to kubernetes could this work without adding more parameters to this docker command?

File: calico_node/allocateipip/allocate_ipip_addr.go [hunk 4]
```diff
@@ -105,40 +130,70 @@ func assignHostTunnelAddr(c *client.Client, nodename string, ipipCidrs []net.IPN
...
+	node, err := c.Nodes().Get(ctx, nodename, options.GetOptions{})
+	if err != nil {
+		log.WithError(err).Fatalf("Unable to retrieve IPIP tunnel address for cleanup. Error getting node '%s'", nodename)
+	}
+
+	node.Spec.BGP.IPv4IPIPTunnelAddr = ipv4Addrs[0].String()
+
+	// If the update fails with ResourceConflict error then retry 5 times with 1 second delay before failing.
+	for i := 0; i < 5; i++ {
+		_, err = c.Nodes().Update(ctx, node, options.SetOptions{})
```
Comment (robbrockbank) [inline calico_node/allocateipip/allocate_ipip_addr.go]: The Get also needs to be part of this loop - otherwise once you hit an update conflict you'll keep hitting it - since it uses the ResourceVersion returned in the Get to decide whether a conflict has occurred or not.

File: calico_node/allocateipip/allocate_ipip_addr.go [hunk 5]
```diff
@@ -105,40 +130,70 @@ func assignHostTunnelAddr(c *client.Client, nodename string, ipipCidrs []net.IPN
...
+			if err != nil {
+				log.WithError(err).WithField("IP", ipv4Addrs[0].String()).Errorf("Error releasing IP address on faiure")
+			}
+
+			log.WithError(err).WithField("IP", ipv4Addrs[0].String()).Fatal("Unable to set IPIP tunnel address")
+		}
+		break
+	}
+
+	log.WithField("IP", ipv4Addrs[0].String()).Info("Set IPIP tunnel address")
```
Comment (robbrockbank) [inline calico_node/allocateipip/allocate_ipip_addr.go]: Not necessarily true - if you hit an error in the loop then you didn't set the tunnel address.

I think you would be better off doing the error check outside this retry loop.   So basically if you hit an error then release the IP address, otherwise log success.

File: calico_node/allocateipip/allocate_ipip_addr.go [hunk 6]
```diff
@@ -105,40 +130,70 @@ func assignHostTunnelAddr(c *client.Client, nodename string, ipipCidrs []net.IPN
...
+	for i := 0; i < 5; i++ {
+		_, err = c.Nodes().Update(ctx, node, options.SetOptions{})
+		if _, ok := err.(cerrors.ErrorResourceUpdateConflict); ok {
+			// Wait for a second and try again if there was a conflict during the resource update.
+			log.Infof("Error updating node %s: %s. Retrying.", node.Name, err)
+			time.Sleep(1 * time.Second)
+			continue
+		} else if err != nil {
+			// We hit an error, so release the IP address before exiting.
+			_, err := c.IPAM().ReleaseIPs(ctx, []net.IP{ipv4Addrs[0]})
```
Comment (robbrockbank) [inline calico_node/allocateipip/allocate_ipip_addr.go]: you don't end up releasing the IP if you have excessive retries.  See comment below.

File: calico_node/allocateipip/allocate_ipip_addr.go [hunk 7]
```diff
@@ -105,40 +130,70 @@ func assignHostTunnelAddr(c *client.Client, nodename string, ipipCidrs []net.IPN
...
+		}
+		break
+	}
+
+	log.WithField("IP", ipv4Addrs[0].String()).Info("Set IPIP tunnel address")
 }
 
 // isIpInPool returns if the IP address is in one of the supplied pools.
-func isIpInPool(ipAddr *net.IP, ipipCidrs []net.IPNet) bool {
+func isIpInPool(ipAddr string, ipipCidrs []net.IPNet) bool {
```
Comment (robbrockbank) [inline calico_node/allocateipip/allocate_ipip_addr.go]: I think the distinction between ipAddr and ipAddress unclear.  Perhaps just name ipAddr as ipAddrStr to make it clear its a string and the other just ipAddr for brevity.

File: calico_node/allocateipip/allocate_ipip_addr.go [hunk 8]
```diff
@@ -35,68 +42,86 @@ func main() {
...
+	}
+
+	ipAddr := net.ParseIP(node.Spec.BGP.IPv4IPIPTunnelAddr)
+	if _, err := c.IPAM().ReleaseIPs(ctx, []net.IP{*ipAddr}); err != nil {
 		log.WithError(err).WithField("IP", ipAddr.String()).Fatal("Error releasing IPIP address from IPAM")
-	} else if err = c.Config().SetNodeIPIPTunnelAddress(nodename, nil); err != nil {
+	}
+
+	node.Spec.BGP.IPv4IPIPTunnelAddr = ""
+	if _, err = c.Nodes().Update(ctx, node, options.SetOptions{}); err != nil {
```
Comment (robbrockbank) [inline calico_node/allocateipip/allocate_ipip_addr.go]: Feels like this should also be encased in a retry loop to handle the conflict error scenario.

File: calico_node/startup/startup.go [hunk 9]
```diff
@@ -552,7 +581,12 @@ func configureIPPools(client *client.Client) {
 	ipv4Present := false
 	ipv6Present := false
 	for _, p := range poolList.Items {
-		version := p.Metadata.CIDR.Version()
+		ip, _, err := net.ParseCIDR(p.Spec.CIDR)
+		if err != nil {
+			fatal("Error parsing CIDR '%s'", p.Spec.CIDR)
+			terminate()
```
Comment (robbrockbank) [inline calico_node/startup/startup.go]: Shouldn't really occur, but perhaps just skip over rather than terminating.

File: calico_node/startup/startup.go [hunk 10]
```diff
@@ -347,19 +370,25 @@ func parseIPEnvironment(envName, envValue string, version int) *net.IPNet {
...
-func validateIP(ipn *net.IPNet) {
+func validateIP(ipn string) {
 	// No validation required if no IP address is specified.
-	if ipn == nil {
+	if ipn == "" {
 		return
 	}
 
+	_, cidr, err := net.ParseCIDR(ipn)
+	if err != nil {
```
Comment (robbrockbank) [inline calico_node/startup/startup.go]: I think it's valid for the IP to just be an IP - in which case we assume a /32 or /128 depending on the version.

So best to use the ParseCIDROrIP method in the net package.

File: calico_node/startup/startup.go [hunk 11]
```diff
@@ -347,19 +370,25 @@ func parseIPEnvironment(envName, envValue string, version int) *net.IPNet {
...
+		fatal("Failed to parse autodetected CIDR '%s'", ipn)
+		terminate()
+	}
+
 	// Pull out the IP as a net.IP (it has useful string and version methods).
-	ip := net.IP{ipn.IP}
+	ip := net.IP{cidr.IP}
```
Comment (robbrockbank) [inline calico_node/startup/startup.go]: I think you just want to use the IP that is returned as the first parameter of the Parse above.  I think the IP for this parsed CIDR will be masked by the network.

File: calico_node/startup/startup.go [hunk 12]
```diff
@@ -616,39 +650,41 @@ func ipv6Supported() bool {
...
-	ipipMode := ipip.Mode(ipipModeName)
-
-	// off is not an actual valid value so switch it to an empty string
-	if ipipModeName == "off" {
-		ipipMode = ipip.Undefined
+	var ipipMode api.IPIPMode
+
+	switch strings.ToLower(ipipModeName) {
+	case "", "off":
+		ipipMode = api.IPIPModeOff
```
Comment (robbrockbank) [inline calico_node/startup/startup.go]: Renamed to "Never".  But we shoul switch on: `"", "off, "never"`

File: calico_node/startup/startup.go [hunk 13]
```diff
@@ -781,83 +855,154 @@ func ensureDefaultConfig(cfg *api.CalicoAPIConfig, c *client.Client, node *api.N
...
-func updateClusterType(c *client.Client, key, clusterType string) error {
+func updateClusterType(ctx context.Context, c client.Interface, clusterType string) error {
 	// If no data is passed in to add, then nothing to do.
 	if clusterType == "" {
 		return nil
 	}
-	global := ""
-	if val, assigned, err := c.Config().GetFelixConfig("ClusterType", global); err != nil {
+
+	// Default ClusterInfo resource should already have been created by this time,
```
Comment (robbrockbank) [inline calico_node/startup/startup.go]: To minimize the number of queries we do on startup we should also be handling the cluster GUID here too since it is in the same resource.

File: calico_node/startup/startup.go [hunk 14]
```diff
@@ -687,35 +723,35 @@ func checkConflictingNodes(client *client.Client, node *api.Node) {
 		// addresses have changed.  If so warn the user as it could be
 		// an indication of multiple nodes using the same name.  This
 		// is not an error condition as the IPs could actually change.
-		if theirNode.Metadata.Name == node.Metadata.Name {
-			if theirIPv4 != nil && ourIPv4 != nil && !theirIPv4.IP.Equal(ourIPv4.IP) {
+		if theirNode.Name == node.Name {
+			if theirIPv4 != "" && ourIPv4 != "" && theirIPv4 != ourIPv4 {
```
Comment (robbrockbank) [inline calico_node/startup/startup.go]: The IP address is actually a CIDR - so we probably need to parse the CIDRs here and compare the IP portion of the CIDR.

File: calico_node/startup/startup.go [hunk 15]
```diff
@@ -781,83 +855,154 @@ func ensureDefaultConfig(cfg *api.CalicoAPIConfig, c *client.Client, node *api.N
...
+				log.WithField("DefaultEndpointToHostAction", felixNodeCfg.Spec.DefaultEndpointToHostAction).Debug("Host Felix value already assigned")
+			}
 		}
 	}
 
 	// Store the Calico Version as a global felix config setting.
-	if err := checkConflictError(c.Config().SetFelixConfig("CalicoVersion", "", VERSION)); err != nil {
-		return err
+	clusterInfo, err := c.ClusterInformation().Get(ctx, GLOBAL_CLUSTER_INFO_NAME, options.GetOptions{})
+	if err != nil {
```
Comment (robbrockbank) [inline calico_node/startup/startup.go]: There are three things we update in the ClusterInformation:  the GUID, the cluster type and the version.   We should do this as part of the same operation otherwise we end up doing more queries/updates than we need to.

File: calico_node/allocateipip/allocate_ipip_addr.go (hunk 4 above)
Comment (gunjan5) [reply calico_node/allocateipip/allocate_ipip_addr.go]: yes yes you're right, I missed that

File: calico_node/startup/startup.go (hunk 14 above)
Comment (gunjan5) [reply calico_node/startup/startup.go]: Not sure I understand the need to parse the CIDR here. If there's a conflict, shouldn't they have the same `IPvXAddress` string with subnet and all?

File: calico_node/startup/startup.go [hunk 16]
```diff
@@ -781,83 +855,154 @@ func ensureDefaultConfig(cfg *api.CalicoAPIConfig, c *client.Client, node *api.N
...
 	// on startup).  Note: KDD does not yet support per-node felix config.
-	if cfg.Spec.DatastoreType != api.Kubernetes {
-		if err := ensureNodeFelixConfig(c, node.Metadata.Name, "DefaultEndpointToHostAction", "RETURN"); err != nil {
-			return err
+	if cfg.Spec.DatastoreType != apiconfig.Kubernetes {
+		felixNodeCfg, err := c.FelixConfigurations().Get(ctx, fmt.Sprintf("%s%s", FELIX_NODE_CONFIG_NAME_PREFIX, node.Name), options.GetOptions{})
+		if err != nil {
+			// Create the default config if it doesn't already exist.
+			if _, ok := err.(cerrors.ErrorResourceDoesNotExist); ok {
+				newFelixNodeCfg := api.NewFelixConfiguration()
```
[the remark this answers is not in the document]
Comment (gunjan5) [reply calico_node/startup/startup.go]: Not sure, I've ported it as it is, but good question. @caseydavenport might know?

[the remark this answers is not in the document]
Comment (caseydavenport) [reply calico_node/startup/startup.go]: I suspect it's because of hybrid-cluster type use-cases, where `calico/node` might set a value that doesn't make sense for say a Felix running on OpenStack or something like that.

Not really sure though, just guessing.

File: calico_node/startup/startup.go (hunk 14 above)
Comment (robbrockbank) [reply calico_node/startup/startup.go]: well theoretically they could have the same IP and different subnet
 -if it's misconfigured.... which is after all what we're testing for

File: calico_node/startup/startup.go (hunk 16 above)
[the remark this answers is not in the document]
Comment (robbrockbank) [reply calico_node/startup/startup.go]: I guess I'm trying to avoid needing per-node configs if at all possible.  In this case perhaps we should just add this to the config file that felix will load if this is a hard requirement.

[the remark this answers is not in the document]
Comment (gunjan5) [reply calico_node/startup/startup.go]: @fasaxc @neiljerram @matthewdupre do you guys know if we need to set `DefaultEndpointToHostAction` to anything other than `RETURN` by default for any of the orchestrators? Can it default to that in Felix? And if other orchestrators need something else then we can set it to other value as part of their setup/manifest

[the remark this answers is not in the document]
Comment (neiljerram) [reply calico_node/startup/startup.go]: I think the Felix default is currently DROP. If we changed that to RETURN, we'd need (if being sufficiently conscientious) OpenStack deployment methods to install different Felix config depending on which version of the Felix code they were installing. Which would not be the end of the world, but another little detail to have to keep track of.

[the remark this answers is not in the document]
Comment (neiljerram) [reply calico_node/startup/startup.go]: Just confirmed that the current Felix default is DROP.

[the remark this answers is not in the document]
Comment (neiljerram) [reply calico_node/startup/startup.go]: On the other hand, I don't _think_ there's any fundamental reason why the default with OpenStack has to be DROP, and it's arguably weird for Calico to default for OpenStack than it does for other orchestrators.

Also, fwiw, haven't yet understood the argument for why, if a different default is needed, it has to be done with per-node config.

[the remark this answers is not in the document]
Comment (neiljerram) [reply calico_node/startup/startup.go]: *to default differently

File: calico_node/allocateipip/allocate_ipip_addr.go (hunk 5 above)
Comment (gunjan5) [reply calico_node/allocateipip/allocate_ipip_addr.go]: if you hit an error (other than updateConflict) we're logging it and exiting with `Fatal()` in line 164

File: calico_node/Makefile (hunk 3 above)
Comment (gunjan5) [reply calico_node/Makefile]: I think ETCD_ENDPOINTS just gets ignored for KDD. 
I will add the other env var needed for KDD as a parameter and set it to it's default value

**Comment by gunjan5** [comment]
CI is failing because the STs are not updated, but the tests for calico-node run successfully `make node-test-containerized`

File: calico_node/startup/startup.go (hunk 16 above)
[the remark this answers is not in the document]
Comment (gunjan5) [reply calico_node/startup/startup.go]: Thanks, @neiljerram that makes sense. 
> I don't think there's any fundamental reason why the default with OpenStack has to be DROP

Des that mean changing the Felix default to `RETURN` wouldn't break OpenStack?  

PS: we set this to `ACCEPT` for [all the k8s manifests](https://docs.projectcalico.org/master/getting-started/kubernetes/installation/hosted/kubeadm/1.6/calico.yaml), so setting this to `RETURN` here doesn't seem to be doing much for k8s.

TL;DR as I understand this:
- Felix defaults to `DROP` (which may or may not be needed for OpenStack)
- Calico-node startup sets it to `RETURN`
- All the k8s manifests then set it to `ACCEPT`

If we're unsure about all the orchestrators then we can leave this as it is for now.

WDYGT? @caseydavenport @robbrockbank @fasaxc @neiljerram

[the remark this answers is not in the document]
Comment (caseydavenport) [reply calico_node/startup/startup.go]: I'll leave this to others to decide - I'm happy either way, but would err on the side of "leave it as it is" since it feels safer and simpler.

[the remark this answers is not in the document]
Comment (neiljerram) [reply calico_node/startup/startup.go]: @gunjan5 I believe it won't fundamentally break OpenStack.  It will hurt release-to-release compatibility of how a Calico/OpenStack deployment behaves - or else add a lot of complexity to what we need to do in all our install methods to ensure that we have consistent behavior.

On balance, I favor preserving the existing behavior.

[the remark this answers is not in the document]
Comment (gunjan5) [reply calico_node/startup/startup.go]: Ok so, in general, we seem to have an agreement not to change the default for now and continue with the existing behavior. I'll keep this behavior as it is in calico-node startup then. Thanks all

File: calico_node/Makefile [hunk 17]
```diff
@@ -498,6 +500,9 @@ node-test-containerized: vendor run-etcd-host run-k8s-apiserver
...
+	-e ETCD_ENDPOINTS=http://$(LOCAL_IP_ENV):2379 \
+	-e KUBECONFIG=$(KUBECONFIG_PATH) \
```
Comment (robbrockbank) [inline calico_node/Makefile]: I might be misreading this, but is this path correct?  This needs to be the path inside the container IIUC and it doesn't look like we are volume mounting that in.

File: calico_node/allocateipip/allocate_ipip_addr.go [hunk 18]
```diff
@@ -2,13 +2,20 @@ package main
 
 import (
 	"os"
+	"time"
 
 	log "github.com/sirupsen/logrus"
 
+	"context"
```
Comment (robbrockbank) [inline calico_node/allocateipip/allocate_ipip_addr.go]: nit:  i think the context import should live with the other native imports at the top.

File: calico_node/startup/startup.go [hunk 19]
```diff
@@ -660,18 +695,27 @@ func createIPPool(client *client.Client, cidr *net.IPNet, ipipModeName string, i
...
+	if nodeList, err := client.Nodes().List(ctx, options.ListOptions{}); err != nil {
 		fatal("Unable to query node confguration: %s", err)
 		terminate()
 	} else {
 		nodes = nodeList.Items
 	}
 
-	ourIPv4 := node.Spec.BGP.IPv4Address
-	ourIPv6 := node.Spec.BGP.IPv6Address
+	ourIPv4, _, err := cnet.ParseCIDR(node.Spec.BGP.IPv4Address)
```
Comment (robbrockbank) [inline calico_node/startup/startup.go]: I thought I'd mentioned this,  I think we should allow just an IP address to be specified too.  So I think you should use the ParseCIDROrIP function - that'll also parse an IP and assume full mask for the network.

File: calico_node/startup/startup.go [hunk 20]
```diff
@@ -781,126 +873,147 @@ func ensureDefaultConfig(cfg *api.CalicoAPIConfig, c *client.Client, node *api.N
 	// for containers to speak to services running on the host (e.g. k8s
 	// pods speaking to k8s api-server, and mesos tasks registering with agent
 	// on startup).  Note: KDD does not yet support per-node felix config.
-	if cfg.Spec.DatastoreType != api.Kubernetes {
-		if err := ensureNodeFelixConfig(c, node.Metadata.Name, "DefaultEndpointToHostAction", "RETURN"); err != nil {
-			return err
+	if cfg.Spec.DatastoreType != apiconfig.Kubernetes {
```
Comment (robbrockbank) [inline calico_node/startup/startup.go]: So was the outcome of the discussions that we always need a per-node config too?  I was rather hoping we wouldn't need that.

File: calico_node/startup/startup_test.go [hunk 21]
```diff
@@ -463,81 +698,82 @@ var _ = Describe("UT for Node IP assignment and conflict checking.", func() {
 	)
 })
```
Comment (robbrockbank) [inline calico_node/startup/startup_test.go]: I believe the latest libcalico-go should support KDD for these types now.

File: calico_node/startup/startup.go (hunk 20 above)
Comment (gunjan5) [reply calico_node/startup/startup.go]: Seems like one reason, as Casey pointed out, was for hybrid deployments where you'd want to have a different per node config for k8s nodes (different from OpenStack)

File: calico_node/startup/startup_test.go (hunk 21 above)
Comment (gunjan5) [reply calico_node/startup/startup_test.go]: ok I pulled in the latest KDD and updated the tests

File: calico_node/startup/startup.go [hunk 22]
```diff
@@ -14,28 +14,34 @@
...
+	"github.com/projectcalico/libcalico-go/lib/options"
+	uuid "github.com/satori/go.uuid"
 	log "github.com/sirupsen/logrus"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
 )
 
 const (
 	DEFAULT_IPV4_POOL_CIDR              = "192.168.0.0/16"
 	DEFAULT_IPV6_POOL_CIDR              = "fd80:24e2:f998:72d6::/64"
+	DEFAULT_IPV4_POOL_NAME              = "default.ipv4.ippool"
```
Comment (gunjan5) [inline calico_node/startup/startup.go]: @robbrockbank lmk if you're ok with these names for the default IPPools

Comment (robbrockbank) [reply calico_node/startup/startup.go]: i think dashes rather than dots, but otherwise fine.

File: calico_node/startup/startup.go [hunk 23]
```diff
@@ -762,17 +813,55 @@ func ensureFilesystemAsExpected() {
...
 	// different from the defaults defined in Felix.
 	//
 	// Logging to file is disabled in the felix.cfg config file.  This
 	// should always be disabled for calico/node.  By default we log to
 	// screen - set the default logging value that we desire.
-	if err := ensureGlobalFelixConfig(c, "ReportingIntervalSecs", "0"); err != nil {
-		return err
-	} else if err = ensureGlobalFelixConfig(c, "LogSeverityScreen", client.GlobalDefaultLogLevel); err != nil {
-		return err
+	felixConf, err := c.FelixConfigurations().Get(ctx, globalFelixConfigName, options.GetOptions{})
```
Comment (robbrockbank) [inline calico_node/startup/startup.go]: I think we should revisit this code later.  When code is this nested it usually means it can be refactored to improve readability.  It is quite difficult to work through.

Let's leave for now, but track a tech debt item to refactor some of this - possibly by introducing some patch mechanism or something.  Not sure.

**Comment by gunjan5** [comment]
Semaphore isn't going to pass because python STs aren't updated, but I ran the node tests locally and they passed, so I'm going to merge this
